packages/dds/map/src/directory.ts

```diff
@@ -1529,8 +1539,27 @@ class SubDirectory extends TypedEventEmitter<IDirectoryEvents> implements IDirec
      * @param op - The message if from a remote delete, or null if from a local delete
      */
     private deleteSubDirectoryCore(subdirName: string, local: boolean) {
-        // This should make the subdirectory structure unreachable so it can be GC'd and won't appear in snapshots
-        // Might want to consider cleaning out the structure more exhaustively though?
-        return this._subdirectories.delete(subdirName);
+        if (this.hasSubDirectory(subdirName)) {
```

Comment:
These events will suffer from the same core issue as #6069, that `needProcessSubDirectoryOperations` (and a couple other locations) has an insufficient design currently (treats resolution algorithm for directory operations as "last-write-wins" when they in fact need something more sophisticated).  As a result neither the eventual consistency nor the events can be trusted until #6069 is resolved.

Because that resolution is more complex than "last-write-wins", we need a different event design.  "valueChanged" and "containedValueChanged" (and their suppression while an op is outstanding) are sufficient for the LWW used on key values, but not good enough for subdirectories where ops prior to the ack of the "last write" still have an impact on the final outcome.  See #8823 for an example where it's less clear what the eventing pattern should be, and some thoughts on design requirements.